Tracking confidence fix

mlurbur opened the pull request:
<!--
Thank you 🙏 for your contribution to [Ultralytics](https://www.ultralytics.com/) 🚀! Your effort in enhancing our repositories is greatly appreciated. To streamline the process and assist us in integrating your Pull Request (PR) effectively, please follow these steps:

1. Check for Existing Contributions: Before submitting, kindly explore existing PRs to ensure your contribution is unique and complementary.
2. Link Related Issues: If your PR addresses an open issue, please link it in your submission. This helps us better understand the context and impact of your contribution.
3. Elaborate Your Changes: Clearly articulate the purpose of your PR. Whether it's a bug fix or a new feature, a detailed description aids in a smoother integration process.
4. Ultralytics Contributor License Agreement (CLA): To uphold the quality and integrity of our project, we require all contributors to sign the CLA. Please confirm your agreement by commenting below:

    I have read the CLA Document and I sign the CLA

For more detailed guidance and best practices on contributing, refer to our ✅ [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Your adherence to these guidelines ensures a faster and more effective review process.
--->
This pull request eliminates misleading and incorrect code behavior for confidence score thresholds during multi object tracking.

Current default tracking behavior:
If the `"conf"` arg is not given, or if it is set equal to zero, it is **overwritten to 0.1 prior to tracking**. This is meant to ensure that low confidence detections are included for ByteTrack.

There are twp issues with the current implementation: 
1. If the user sets conf to 0.0, it will be **silently overwritten**, because `kwargs.get("conf")` evaluates to False when "conf" == 0.
2. If a user adjusts the confidence threshold of ByteTrack or any other tracker to be lower than 0.1, without also explicitly setting the confidence of the model lower than that value, the hard-coded confidence threshold will override that argument, **removing expected detections** without the user knowing. This contradicts the implemented design philosophy, which is to ensure no detections are witheld from the tracker.

I have read the CLA Document and I sign the CLA


## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Improves tracking reliability in YOLO by warning users when their detection confidence threshold may cause tracking issues, and updates documentation and tests for clarity and robustness. 🚦📚

### 📊 Key Changes
- **User Warning Added:** Now warns users if the detection confidence threshold is set higher than the recommended value for ByteTrack or BoTSORT trackers, helping prevent missed detections during tracking.
- **Test Coverage Improved:** Adds a test to ensure the warning is triggered correctly.
- **Code Clarity:** Refines how the confidence threshold is set in the model code for better reliability.
- **Documentation Updates:** Minor improvements to code examples in the k-fold cross-validation and Triton Inference Server guides for better readability and accuracy.

### 🎯 Purpose & Impact
- **Better User Guidance:** Users are now alerted if their settings might reduce tracking performance, making it easier to achieve optimal results.
- **Increased Reliability:** Helps prevent common mistakes that could lead to poor tracking, especially for those using ByteTrack or BoTSORT.
- **Clearer Docs:** Documentation tweaks make it easier for users to follow best practices and set up their workflows smoothly.
- **Improved Testing:** Ensures new warning behavior is reliable and won’t break in future updates.

Overall, these changes make tracking with YOLO models more user-friendly and robust, especially for those working with video analytics or multi-object tracking tasks.

mlurbur committed:
Update model.py

mlurbur committed:
Update model.py

UltralyticsAssistant commented:
👋 Hello @mlurbur, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Thank you for clearly explaining the purpose of your fix in your PR description. Linking to any [related issues](https://github.com/ultralytics/ultralytics/issues) would further help contextualize your changes.  
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.  
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. CI results are critical to evaluate whether your code is functioning as expected. Please address any failing checks if there are any.  
- ✅ **Update Documentation**: If applicable, update the relevant [documentation](https://docs.ultralytics.com/) to reflect changes made in this PR. Changes related to configurations like confidence thresholds may benefit from added clarity in the documentation.  
- ✅ **Add Tests**: If your PR involves behavior changes or bug fixes, please consider adding or updating unit tests to ensure coverage of the new logic. Confirm that all tests are passing as part of your submission.  
- ✅ **Sign the CLA**: Thank you for signing the [Contributor License Agreement (CLA)](https://docs.ultralytics.com/help/CLA/)!  

### Additional Suggestions 🛠️  
- **Minimize Changes**: Limit your changes to the minimum necessary for the fix. While your description highlights the reasoning behind the adjustments, make sure the implementation is concise and in line with the existing codebase. 🌟  
- **Testing with an MRE**: For bug fixes, providing a Minimum Reproducible Example (MRE) allows us to verify the issue and validate the fix. If not already included, consider adding an example showcasing the behavior of threshold changes on tracking performance.  

### Next Steps ✅  
- An Ultralytics engineer will review your PR soon to assist with further feedback or merging it into the repository! 💬  

For additional information, refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don’t hesitate to comment below if you require further clarification. Thank you for contributing to Ultralytics! 🚀

mlurbur committed:
Update model.py

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com

glenn-jocher commented:
@Laughing-q can you please take a look at this PR?

Laughing-q committed:
Merge branch 'main' into tracking_confidence_fix

Laughing-q commented:
@mlurbur Hi thanks for the suggestions here! The `0.1` here is from the recommended setting from original bytetrack paper, where they defined the low-scores ranges from `0.1` to `0.4`. Also generally predictions in one image could a mess when we set extremely low score threshold, I think we better keep 0.1 as lowest default value.
May I ask why you want to set threshold to 0.0? Thanks!

mlurbur commented:
Hi @Laughing-q, thanks for your response. 

My reasoning is as follows:
The "default" detection confidence for the model when tracking should never be greater than the `track_low_thresh` from ByteTrack `.yaml` settings (or any min confidence threshold of future trackers you might add). In the current implementation, if I edit the `track_low_thresh` in `[bytetrack.yaml](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/cfg/trackers/bytetrack.yaml)` to 0.01 and don't also explicitly set the `conf` of the model to 0.01, the default behavior of tracking is incorrect. Expected detections will be filtered out after inference because of the hard-coded value of 0.1

To your point about affecting user experience:
This should not degrade user experience because all detections with a confidence below the `track_low_thresh` are removed during tracking (see lines 311-317 of `[byte_tracker.py](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/trackers/byte_tracker.py)`).

You can run this code to test, which shows the results are identical when conf=0 or when default behavior is used, even though there are a different number of detections.

```
from ultralytics import YOLO

video_url = "https://github.com/ultralytics/assets/releases/download/v0.0.0/decelera_portrait_min.mov"
model = YOLO("yolo11n.pt")
tracker = "bytetrack.yaml"
# showing that when conf=0.0 there are more detections
default_detection_results = model.predict(source=video_url, imgsz=160)
zero_conf_detection_results = model.predict(source=video_url, imgsz=160, conf=0.0)

# test that model produces more detections when conf=0.0
for dp_r, zp_r in zip(default_detection_results, zero_conf_detection_results):
    assert len(zp_r) > len(dp_r)

# tracking with default behavior and with conf=0.0
default_tracking_results = model.track(source=video_url, imgsz=160, tracker=tracker)
zero_conf_tracking_results = model.track(source=video_url, imgsz=160, tracker=tracker, conf=0.0)

# test that tracking still gives identical results
for d_r, z_r in zip(default_tracking_results, zero_conf_tracking_results):
    assert d_r.boxes.data.equal(z_r.boxes.data)
```

Laughing-q commented:
@mlurbur yeah I think you are right for track, as long as we have a decent `track_low_thresh` variable there. However using `conf=0.0` by default would slow the postprocessing(nms) 2x, in this case every user would be affected by this 2x slow down.
```bash
yolo detect predict source=/home/laughing/codes/datasets/coco128/images/train2017  conf=0.1
Speed: 0.7ms preprocess, 7.4ms inference, 0.8ms postprocess per image at shape (1, 3, 448, 640)
```
```bash
yolo detect predict source=/home/laughing/codes/datasets/coco128/images/train2017  conf=0.0
Speed: 0.7ms preprocess, 7.4ms inference, 2.3ms postprocess per image at shape (1, 3, 448, 640)
```
see the postprocessing time here.
I understand your point of the needs of tuning the thresholds of tracker, I think the best solution here is to add a warning when the passed `track_low_thresh` is lower than the `conf` we are using, notify users to modify the `conf` as well.

Laughing-q committed:
Merge branch 'main' into tracking_confidence_fix

RizwanMunawar committed:
Merge branch 'main' into tracking_confidence_fix

mlurbur committed:
add check and warning message

mlurbur committed:
Revert to 0.1 conf. Retain None logic

mlurbur committed:
add testing for edge case and warning

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com

mlurbur commented:
@Laughing-q I added checking logic and a warning in `track.py` to catch the bug. I've reverted the default tracking confidence to 0.1 but retained the `if kwargs.get("conf") is not None else 0.1` logic in `model.py` which ensures that if a user sets `conf=0.0` that will not be overwritten as 0.1.

Also added a test for the additional check.

The auto format bot created merge conflicts in a random documentation file fyi
